packages/tailwindcss/src/plugin-api.ts

```diff
@@ -96,6 +103,71 @@ function buildPluginApi(
         designSystem.variants.fromAst(name, objectToAst(variant))
       }
     },
+    matchVariant(name, fn, options) {
+      function resolveVariantValue<T extends Parameters<typeof fn>[0]>(
+        value: T,
+        modifier: CandidateModifier | null,
+        nodes: AstNode[],
+      ): AstNode[] {
+        let resolved = fn(value, { modifier: modifier?.value ?? null })
+        return parseVariantValue(resolved, nodes)
+      }
+
+      let defaultOptionKeys = Object.keys(options?.values ?? {})
+      designSystem.variants.group(
+        () => {
+          designSystem.variants.functional(name, (ruleNodes, variant) => {
+            if (!variant.value || variant.modifier) {
+              if (options?.values && 'DEFAULT' in options.values) {
+                ruleNodes.nodes = resolveVariantValue(options.values.DEFAULT, null, ruleNodes.nodes)
+                return
+              }
+              return null
+            }
+
+            if (variant.value.kind === 'arbitrary') {
+              ruleNodes.nodes = resolveVariantValue(
+                variant.value.value,
+                variant.modifier,
+                ruleNodes.nodes,
+              )
+            } else if (variant.value.kind === 'named' && options?.values) {
+              let defaultValue = options.values[variant.value.value]
+              if (typeof defaultValue !== 'string') {
+                return
+              }
+
+              ruleNodes.nodes = resolveVariantValue(defaultValue, null, ruleNodes.nodes)
+            }
+          })
+        },
+        (a, z) => {
+          // Since we only define a functional variant in the group, the `kind`
+          // has to be `functional`.
+          if (a.kind !== 'functional' || z.kind !== 'functional') {
+            return 0
+          }
+          if (!a.value || !z.value) {
+            return 0
+          }
+
+          if (options && typeof options.sort === 'function') {
+            let aValue = options?.values?.[a.value.value] ?? a.value.value
+            let zValue = options?.values?.[z.value.value] ?? z.value.value
+
+            return options.sort(
+              { value: aValue, modifier: a.modifier?.value ?? null },
+              { value: zValue, modifier: z.modifier?.value ?? null },
+            )
+          }
+
+          let aOrder = defaultOptionKeys.indexOf(a.value.value) ?? 0
+          let zOrder = defaultOptionKeys.indexOf(z.value.value) ?? 0
```

Comment:
Don't think we will ever hit the `?? 0` branch since `indexOf(…)` would return `-1`. Man, I wish we had proper enums in JS 😂

```suggestion
          let aOrder = defaultOptionKeys.indexOf(a.value.value)
          let zOrder = defaultOptionKeys.indexOf(z.value.value)
```